File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int n = 0;
        stack<TreeNode*> st;
        TreeNode* node = root;
        
        while(node != NULL || !st.empty()){
            while(node != NULL){
                st.push(node);
                node = node->left;
            }
            node = st.top();
            st.pop();
            n++;
            if(n==k){
                return node->val;
            }
            node = node->right;
        }
    }

    // int kthSmallest(TreeNode* root, int k) {
    //     vector<int> arr;
    //     dfs(root, arr);
    //     return arr[k-1];
    // }

    // void dfs(TreeNode* root, vector<int>& arr){
    //     if(root==NULL) return;
    //     dfs(root->left, arr);
    //     arr.push_back(root->val);
    //     dfs(root->right, arr);
    // }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        self.res = 0
        if root is None:
            return 0
        def dfs(root, maxValue):
            if root is None:
                return
            if root.val >= maxValue:
                self.res+=1
            maxValue = max(root.val, maxValue)
            dfs(root.left, maxValue)
            dfs(root.right, maxValue)
            return

        n=float("-inf")
        dfs(root, n)

        return self.res

    

